kernel/safe_typing.ml

```diff
@@ -790,7 +785,20 @@ let add_constant ?role ~in_section l decl senv =
   let eff = match role with
   | None -> empty_private_constants
   | Some role ->
-    let cb = map_constant Future.force cb in
+    let body, univs = match cb.const_body with
+    | (Primitive _ | Undef _) -> assert false
+    | Def c -> (Def c, cb.const_universes)
+    | OpaqueDef o ->
+      let (b, ctx) = Future.force o in
+      match cb.const_universes with
+      | Monomorphic ctx' ->
+        OpaqueDef b, Monomorphic (Univ.ContextSet.union ctx ctx')
+      | Polymorphic auctx ->
+        (* Upper layers enforce that these constraints are not delayed *)
```

Comment:
Shouldn't this rather be "upper layers enforce that there are no private universes in side effects" or some such?